plugin/helpers_bots.go

```diff
@@ -72,3 +113,67 @@ func (p *HelpersImpl) EnsureBot(bot *model.Bot) (retBotId string, retErr error)
 
 	return createdBot.UserId, nil
 }
+
+func (p *HelpersImpl) ShouldProcessMessage(post *model.Post, botUserId string, options ...ShouldProcessMessageOption) bool {
+	messageProcessOptions := initialiseOptions(options)
+
+	if !messageProcessOptions.AllowSelf {
+		if post.UserId == botUserId {
+			return false
+		}
+	}
+
+	if !messageProcessOptions.AllowSystemMessages {
+		if post.IsSystemMessage() {
+			return false
+		}
+	}
+
+	if !contains(messageProcessOptions.FilterChannelIDs, post.ChannelId) {
+		channel, _ := p.API.GetChannel(post.ChannelId)
+
+		if !IsBotDMChannel(channel, botUserId) {
+			return false
+		}
+	}
+
+	if !contains(messageProcessOptions.FilterUserIDs, post.UserId) {
+		user, _ := p.API.GetUser(post.UserId)
+		if !messageProcessOptions.AllowBots {
+			if user.IsBot {
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func initialiseOptions(options []ShouldProcessMessageOption) *shouldProcessMessageOptions {
+	messageProcessOptions := &shouldProcessMessageOptions{}
+	for _, option := range options {
+		option(messageProcessOptions)
+	}
+	return messageProcessOptions
+}
+
+func IsBotDMChannel(channel *model.Channel, botUserID string) bool {
+	if channel.Type != model.CHANNEL_DIRECT {
+		return false
+	}
+
+	if !strings.HasPrefix(channel.Name, botUserID+"__") && !strings.HasSuffix(channel.Name, "__"+botUserID) {
+		return false
+	}
+
+	return true
+}
+
+func contains(a []string, x string) bool {
```

Comment:
You can use `utils.StringInSlice` instead

Reply:
Done